### ISSUE TITLE
the route of yolov4-tiny.cfg

### ISSUE THREAD

**Issue by XDhughie** [AlexeyAB/darknet#6086]
[route]
layers=-1
groups=2
group_id=1

what doed the "groups=2 group_id=1" mean?

**Comment by WongKinYiu** [comment]
split previous layer (layers=-1) into two parts through channel (groups=2) and route the second part (group_id=1, id start from 0).

**Comment by ai815** [comment]
> 
> 
> split previous layer (layers=-1) into two parts through channel (groups=2) and route the second part (group_id=1, id start from 0).

hey, can I ask what the meaning of channel  here?

**Comment by WongKinYiu** [comment]
size of feature map is width * height * channel.

**Comment by ai815** [comment]
> 
> 
> size of feature map is width * height * channel.

oh  I see thanks. but how could we seperaste a layer into two parts, can you plz explain?

**Comment by WongKinYiu** [comment]
if width * height * channel is w * h * c, the feature map will be fm[0:w, 0:h, 0:c].
groups = 2, group_id = 0 will gets fm[0:w, 0:h, 0:c/2].
groups = 2, group_id = 1 will gets fm[0:w, 0:h, c/2:c].

**Comment by ankandrew** [comment]
@WongKinYiu I don't understand why (visually with Netron) the num. of c is 64

![image](https://user-images.githubusercontent.com/61120139/86020771-b48a0f80-b9fe-11ea-9b38-37825f7325e4.png)


Shouldn't be 32 because `c/2`?

**Comment by WongKinYiu** [comment]
maybe it because darknet use group_id but netron use groups_id.
https://github.com/AlexeyAB/darknet/blob/master/src/parser.c#L1036
https://github.com/lutzroeder/netron/blob/master/src/darknet-metadata.json#L281

you can try to modify the code of netron and generate the graph again.

**Comment by ankandrew** [comment]
@WongKinYiu I just tried changing `group_id` to `groups_id` but they produce both the same result. I should open an issue in netron I guess

**Comment by ShaneHsieh** [comment]
@WongKinYiu Hi

Yolov4-tiny only use the group_id=1? So many channel is thrown away?

**Comment by WongKinYiu** [comment]
no, cross stage connection routes all of channel of base layer in yolov4-tiny.

**Comment by ShaneHsieh** [comment]
@WongKinYiu I can understand it.
You mean that only this route layer uses half channel,but previous layer that uses all channel will be concat?

**Comment by WongKinYiu** [comment]
![image](https://user-images.githubusercontent.com/12152972/86105021-6570d700-baf1-11ea-9f19-b2ca3f644afb.png)

**Comment by hhaAndroid** [comment]
is right?

`class ResConv2dBatchLeaky(nn.Module):

    def __init__(self, in_channels, inter_channels, kernel_size, stride=1, leaky_slope=0.1, return_extra=False):
        super(ResConv2dBatchLeaky, self).__init__()

        self.return_extra = return_extra
        self.in_channels = in_channels
        self.inter_channels = inter_channels
        self.kernel_size = kernel_size
        self.stride = stride
        if isinstance(kernel_size, (list, tuple)):
            self.padding = [int(ii / 2) for ii in kernel_size]
        else:
            self.padding = int(kernel_size / 2)
        self.leaky_slope = leaky_slope

        self.layers0 = Conv2dBatchLeaky(self.in_channels//2, self.inter_channels, self.kernel_size, self.stride,
                                        self.padding)
        self.layers1 = Conv2dBatchLeaky(self.inter_channels, self.inter_channels, self.kernel_size, self.stride,
                                        self.padding)
        self.layers2 = Conv2dBatchLeaky(self.in_channels, self.in_channels, 1, 1, 0)

    def forward(self, x):
        y0 = x
        channel = x.shape[1]
        x0 = x[:, channel // 2:, ...]
        x1 = self.layers0(x0)
        x2 = self.layers1(x1)
        x3 = torch.cat((x2, x1), dim=1)
        x4 = self.layers2(x3)
        x = torch.cat((y0, x4), dim=1)
        if self.return_extra:
            return x, x4
        else:
            return x`

**Comment by hhaAndroid** [comment]
The whole model structure is as follows? is right?

`class TinyYolov4(nn.Module):

    def __init__(self, pretrained=False):
        super(TinyYolov4, self).__init__()

        # Network
        backbone = OrderedDict([
            ('0_convbatch', vn_layer.Conv2dBatchLeaky(3, 32, 3, 2)),
            ('1_convbatch', vn_layer.Conv2dBatchLeaky(32, 64, 3, 2)),
            ('2_convbatch', vn_layer.Conv2dBatchLeaky(64, 64, 3, 1)),
            ('3_resconvbatch', vn_layer.ResConv2dBatchLeaky(64, 32, 3, 1)),
            ('4_max', nn.MaxPool2d(2, 2)),
            ('5_convbatch', vn_layer.Conv2dBatchLeaky(128, 128, 3, 1)),
            ('6_resconvbatch', vn_layer.ResConv2dBatchLeaky(128, 64, 3, 1)),
            ('7_max', nn.MaxPool2d(2, 2)),
            ('8_convbatch', vn_layer.Conv2dBatchLeaky(256, 256, 3, 1)),
            ('9_resconvbatch', vn_layer.ResConv2dBatchLeaky(256, 128, 3, 1, return_extra=True)),
        ])

        head = [
            OrderedDict([
                ('10_max', nn.MaxPool2d(2, 2)),
                ('11_conv', vn_layer.Conv2dBatchLeaky(512, 512, 3, 1)),
                ('12_conv', vn_layer.Conv2dBatchLeaky(512, 256, 1, 1)),
            ]),

            OrderedDict([
                ('13_conv', vn_layer.Conv2dBatchLeaky(256, 512, 3, 1)),
                ('14_conv', nn.Conv2d(512, 3 * (5 + 80), 1)),
            ]),

            OrderedDict([
                ('15_convbatch', vn_layer.Conv2dBatchLeaky(256, 128, 1, 1)),
                ('16_upsample', nn.Upsample(scale_factor=2)),
            ]),

            OrderedDict([
                ('17_convbatch', vn_layer.Conv2dBatchLeaky(384, 256, 3, 1)),
                ('18_conv', nn.Conv2d(256, 3 * (5 + 80), 1)),
            ]),
        ]

        self.backbone = nn.Sequential(backbone)
        self.head = nn.ModuleList([nn.Sequential(layer_dict) for layer_dict in head])
        self.init_weights(pretrained)

    def forward(self, x):
        stem, extra_x = self.backbone(x)
        stage0 = self.head[0](stem)
        head0 = self.head[1](stage0)

        stage1 = self.head[2](stage0)
        stage2 = torch.cat((stage1, extra_x), dim=1)
        head1 = self.head[3](stage2)
        head = [head1, head0]
        return head`

**Comment by WongKinYiu** [comment]
i think yes.

```python
class ResConv2dBatchLeaky(nn.Module):
  def __init__(self, in_channels, inter_channels, kernel_size, stride=1, leaky_slope=0.1, return_extra=False):
    super(ResConv2dBatchLeaky, self).__init__()

    self.return_extra = return_extra
    self.in_channels = in_channels
    self.inter_channels = inter_channels
    self.kernel_size = kernel_size
    self.stride = stride
    if isinstance(kernel_size, (list, tuple)):
        self.padding = [int(ii / 2) for ii in kernel_size]
    else:
        self.padding = int(kernel_size / 2)
    self.leaky_slope = leaky_slope

    self.layers0 = Conv2dBatchLeaky(self.in_channels//2, self.inter_channels, self.kernel_size, self.stride,
                                    self.padding)
    self.layers1 = Conv2dBatchLeaky(self.inter_channels, self.inter_channels, self.kernel_size, self.stride,
                                    self.padding)
    self.layers2 = Conv2dBatchLeaky(self.in_channels, self.in_channels, 1, 1, 0)

  def forward(self, x):
    y0 = x
    channel = x.shape[1]
    x0 = x[:, channel // 2:, ...]
    x1 = self.layers0(x0)
    x2 = self.layers1(x1)
    x3 = torch.cat((x2, x1), dim=1)
    x4 = self.layers2(x3)
    x = torch.cat((y0, x4), dim=1)
    if self.return_extra:
        return x, x4
    else:
        return x
```

**Comment by willbattel** [comment]
How come v4 tiny uses the route groups in its cfg file, but the full v4 cfg does not? It looks like they're both CSP-based, but I'm not sure why only v4 tiny uses the route groups. Is it just to increase speed by cutting the feature size in half?

**Comment by WongKinYiu** [comment]
`groups` and `group_id` are implemented after cspnet is designed, so new csp models such as yolov4-tiny implement csp using route groups.

**Comment by willbattel** [comment]
Interesting. So if YOLOv4 had not been published until later, it would have used the route groups? And I assume the functionality would have been the same as it is currently, just a different implementation? Thanks.

**Comment by WongKinYiu** [comment]
yes, both of two implementations are equivalent.

**Comment by willbattel** [comment]
I'm reading the CSPNet paper again, and I'm noticing some discrepancies in how CSPDarknet53 adopted this mechanism.

In CSPNet, the base layer is split in two. One half goes into the block, and the other half is concatenated with the output of the block.

![csp](https://user-images.githubusercontent.com/13490339/88341746-c6ac5480-cd03-11ea-826b-14981fa438c0.PNG)

Maybe I'm interpreting the Darknet code wrong, but it looks like CSPDarknet53 does something a bit different. Like CSPNet, CSPDarknet53 splits the base layer in two, and then sends one half into the block. However, then it looks like the _entire_ base layer- **not just the other half**- is concatenated with the block output.

In `yolov4-tiny.cfg`, the `route` layer with the `groups` and `group_id` properties send forward only half of the channels to the following layers. However, then it looks like the entire base layer is concatenated to the block output- not just the other half. https://github.com/AlexeyAB/darknet/blob/master/cfg/yolov4-tiny.cfg#L81

Also, in `yolov4.cfg`, I was having difficulty determining which layers were getting half of the feature map, and which were getting the full feature map.

I was also looking through some TensorFlow adaptions of CSPDarknet53, and of the three I looked at: all had different configurations.

This repo performs like CSPNet, where one half goes through the block and other other half is concatenated: https://github.com/sicara/tf2-yolov4/blob/master/tf2_yolov4/backbones/csp_darknet53.py#L30

This repo appears to concatenate the entire base layer onto the block output: https://github.com/hunglc007/tensorflow-yolov4-tflite/blob/master/core/backbone.py#L51

And this repo seems to mimick `yolov4.cfg`, where the filters are manually halved: https://github.com/RobotEdh/Yolov-4/blob/master/YoloV4_tf.ipynb

So which two of these four patterns (one pair for full and one pair for tiny) are correct?

![yolocsp](https://user-images.githubusercontent.com/13490339/88343651-7d5e0400-cd07-11ea-9123-c20b3a42ad95.PNG)

1. Does YOLOv4 concatenate the entire base layer (B1), or only half of the base layer (A1), to the output of the block?
2. Does YOLOv4 Tiny concatenate the entire base layer (B2), or only half of the base layer (A2), to the output of the block?

**Comment by WongKinYiu** [comment]
yolov4 is based on resnet, it split channel in base layer and remove bottleneck of res layers. so two path will be {2x, 2x}.
yolov4-tiny is based on vovnet, if we split channel in base layer too, two path will be {1x, 3x}. so we split it in computational block to make it be {2x, 2x}. this modification is used for optimizing memory bandwidth.

**Comment by Lowell-IC** [comment]
> if width * height * channel is w * h * c, the feature map will be fm[0:w-1, 0:h-1, 0:c-1].
> groups = 2, group_id = 0 will gets fm[0:w-1, 0:h-1, 0:c/2-1].
> groups = 2, group_id = 1 will gets fm[0:w-1, 0:h-1, c/2:c-1].
